mpisppy/opt/presolve.py

```diff
@@ -0,0 +1,247 @@
+
+import abc
+import weakref
+
+import numpy as np
+
+from pyomo.common.errors import InfeasibleConstraintException, DeferredImportError
+from pyomo.contrib.appsi.fbbt import IntervalTightener
+
+from mpisppy import MPI
+from mpisppy import global_toc
+
+
+class _SPPresolver(abc.ABC):
+    """Defines a presolver for distributed stochastic optimization problems
+
+    Args:
+        spbase (SPBase): an SPBase object
+    """
+
+    def __init__(self, spbase):
+        self._opt = None
+        self.opt = spbase
+
+    @abc.abstractmethod
+    def presolve(self) -> bool:
+        """should return `True` if modifications are made, `False` otherwise"""
+
+    @property
+    def opt(self):
+        if self._opt is None:
+            return None
+        return self._opt()
+
+    @opt.setter
+    def opt(self, value):
+        if self._opt is None:
+            self._opt = weakref.ref(value)
+        else:
+            raise RuntimeError("SPPresolve.opt should only be set once")
+
+
+class SPIntervalTightener(_SPPresolver):
+    """Interval Tightener (feasibility-based bounds tightening)
+    TODO: enable options
+    """
+
+    def __init__(self, spbase):
+        super().__init__(spbase)
+
+        self.subproblem_tighteners = {}
+        for k, s in self.opt.local_subproblems.items():
+            try:
+                self.subproblem_tighteners[k] = it = IntervalTightener()
+            except DeferredImportError as e:
+                # User may not have extension built
+                raise ImportError(f"presolve needs the APPSI extensions for pyomo: {e}")
+
+            # ideally, we'd be able to share the `_cmodel`
+            # here between interfaces, etc.
+            try:
+                it.set_instance(s)
+            except Exception as e:
+                # TODO: IntervalTightener won't handle
+                # every Pyomo model smoothly, see:
+                # https://github.com/Pyomo/pyomo/issues/3002
+                # https://github.com/Pyomo/pyomo/issues/3184
+                # https://github.com/Pyomo/pyomo/issues/1864#issuecomment-1989164335
+                raise Exception(
+                    f"Issue with IntervalTightener; cannot apply presolve to this problem. Error: {e}"
+                )
+
+    def presolve(self):
+        """Run the interval tightener (FBBT):
+        1. FBBT on each subproblem
+        2. Narrow bounds on the nonants across all subproblems
+        3. If the bounds are updated, go to (1)
+        """
+
+        update = False
+
+        while True:
+            big_iters = 0.0
+            for k, it in self.subproblem_tighteners.items():
+                n_iters = it.perform_fbbt(self.opt.local_subproblems[k])
+                # get the number of constraints after we do
+                # FBBT so we get any updates on the subproblem
+                big_iters = max(big_iters, n_iters / len(it._cmodel.constraints))
+
+            update_this_pass = big_iters > 1.0
```

Comment:
This is saying, if any variables were tightened, keep going. Otherwise break. Right?